src/chunk.c

```diff
@@ -314,42 +404,45 @@ chunk_create_after_lock(Hypertable *ht, Point *p, const char *schema, const char
 	chunk_collision_resolve(hs, cube, p);
 
 	/* Create a new chunk based on the hypercube */
+	catalog_become_owner(catalog, &sec_ctx);
 	chunk = chunk_create_stub(catalog_table_next_seq_id(catalog, CHUNK), hs->num_dimensions);
+	catalog_restore_user(&sec_ctx);
+
 	chunk->fd.hypertable_id = hs->hypertable_id;
 	chunk->cube = cube;
-	chunk->num_constraints = chunk->capacity;
+	chunk->hypertable_relid = ht->main_table_relid;
 	namestrcpy(&chunk->fd.schema_name, schema);
-
 	snprintf(chunk->fd.table_name.data, NAMEDATALEN,
 			 "%s_%d_chunk", prefix, chunk->fd.id);
 
+	/* Insert chunk. Chunk table already locked in exclusive mode. */
+	chunk_insert_lock(chunk, NoLock);
+
 	/* Insert any new dimension slices */
 	dimension_slice_insert_multi(cube->slices, cube->num_slices);
 
-	/* All slices now have assigned ID's so update the chunk's constraints */
-	for (i = 0; i < hs->num_dimensions; i++)
-	{
-		chunk->constraints[i].fd.chunk_id = chunk->fd.id;
-		chunk->constraints[i].fd.dimension_slice_id = cube->slices[i]->fd.id;
-	}
-
-	/* Insert the new chunk constraints */
-	chunk_constraint_insert_multi(chunk->constraints, chunk->num_constraints);
-	process_utility_set_expect_chunk_modification(true);
-	CatalogInternalCall4(CHUNK_CREATE, Int32GetDatum(chunk->fd.id), Int32GetDatum(hs->hypertable_id),
-			   CStringGetDatum(schema), NameGetDatum(&chunk->fd.table_name));
-	process_utility_set_expect_chunk_modification(false);
+	/* Add metadata for dimensional and inheritable constraints */
+	chunk_add_constraints(chunk);
 
-	chunk->table_id = get_relname_relid(NameStr(chunk->fd.table_name), schema_oid);
+	/* Create the actual table relation for the chunk */
+	chunk->table_id = chunk_create_table(chunk, ht);
 
-	trigger_create_all_on_chunk(ht, chunk);
+	if (!OidIsValid(chunk->table_id))
+		elog(ERROR, "Could not create chunk table");
 
-	/* Create all indexes on the chunk */
-	chunk_index_create_all(ht->fd.id, ht->main_table_relid, chunk->fd.id, chunk->table_id);
+	/* Create the chunk's constraints, triggers, and indexes */
+	chunk_constraints_create(chunk->constraints,
```

Comment:
Previously, we created "dimensional" constraints separately from "inherited" constraints. Dimensional constraints were added when the chunk table was created (basically here), while inherited constraints were added after inherited triggers and indexes. Now all constraints, no matter the type, are added after table creation. If you believe it matters (I don't see why it would myself), I can move the constraints to the end. But that will instead change the order of the dimensional constraints.

Reply:
Ah ok, that makes sense.